src/http-api.js

```diff
@@ -106,6 +106,136 @@ module.exports.MatrixHttpApi.prototype = {
+        if (rawResponse === undefined) {
+            if (global.XMLHttpRequest) {
+                rawResponse = false;
+            } else {
+                console.warn(
+                    "Returning the raw JSON from uploadContent(). Future " +
+                    "versions of the js-sdk will change this default, to " +
+                    "return the parsed object. Set opts.rawResponse=false " +
+                    "to change this behaviour now.");
+                rawResponse = true;
+            }
+        }
+
+        let promise;
+
+        // XMLHttpRequest doesn't parse JSON for us. request normally does, but
+        // we're setting opts.json=false so that it doesn't JSON-encode the
+        // request, which also means it doesn't JSON-decode the response. Either
+        // way, we have to JSON-parse the response ourselves.
+        let bodyParser = null;
+        if (!rawResponse) {
+            bodyParser = function(rawBody) {
+                return JSON.parse(rawBody);
+            };
+        }
+
+        if (global.XMLHttpRequest) {
+            const defer = Promise.defer();
+            const xhr = new global.XMLHttpRequest();
+            const cb = requestCallback(defer, opts.callback, this.opts.onlyData);
+
+            const timeout_fn = function() {
+                xhr.abort();
+                cb(new Error('Timeout'));
+            };
+
+            // set an initial timeout of 30s; we'll advance it each time we get
+            // a progress notification
+            xhr.timeout_timer = callbacks.setTimeout(timeout_fn, 30000);
+
+            xhr.onreadystatechange = function() {
+                switch (xhr.readyState) {
+                    case global.XMLHttpRequest.DONE:
+                        callbacks.clearTimeout(xhr.timeout_timer);
+                        let resp;
+                        try {
+                            if (!xhr.responseText) {
+                                throw new Error('No response body.');
+                            }
+
+                            resp = xhr.responseText;
+                            if (bodyParser) {
+                                resp = bodyParser(resp);
+                            }
+
+                            if (xhr.status != 200) {
+                                const { error } = resp;
+                                cb({ message: error, http_status: xhr.status });
+                                return;
+                            }
+                        } catch (err) {
+                            err.http_status = xhr.status;
+                            cb(err);
+                            return;
+                        }
+                        cb(undefined, xhr, resp);
+                        break;
+                }
+            };
+            let url = this.opts.baseUrl + "/_matrix/media/v1/resolve_url";
+            url += "?access_token=" + encodeURIComponent(this.opts.accessToken);
```

Comment:
if `this.useAuthorizationHeader` is true we send the access token via "Authorization Bearer <auth_token" instead of query param.